docs/website/contents/for-developers/HardForkCombinatorRollbackMonotonicityProposal.md

```diff
@@ -0,0 +1,190 @@
+## Though Experiment: Predictions within a Simplified Node
+
+Consider a simplified node.
+
+- Only node-to-node protocols: no queries, etc.
+- Pure Praos: no hard fork combinator, no Transitional Praos, etc.
+
+### Question 1
+
+Which aspects of the protocol and ledger's future behavior must this node be able to predict from a given ledger state and why?
+
+- Assume it's well-connected and running in a healthy net.
+- Assume k > 0.
+- Let D be the node's currently selected chain.
+- Let C be the current best valid chain in the net.
+- Let L be the ledger state resulting from the youngest block that is on both C and D.
+- Let R count the number of blocks after L on D.
+  Observe that there must be more than R blocks after L on C, since C is better than D.
+  (This isn't necessarily true, due to tiebreakers.
+  However, even with tiebreakers, this remains true in the worst case, which is what ultimately matters.)
+- Common Prefix ensures R <= k.
+- L must contain enough information for the node to correctly judge the validity of the k+1th block after L on C.
+- Otherwise the Header-Body Split would prevent the node from adopting C, even though doing so requires at most rolling back k blocks.
+
+That is a complete answer.
+In particular, there is nothing else this simplified node needs to predict.
+
+If we assume Chain Growth in addition to Common Prefix, then we can further refine the answer.
+
+- Chain Growth ensures that every window of scg slots will contain at least k blocks on C.
+- Because of the Header-Body Split, L must contain enough information for the node to correctly judge the validity of any header that is no more than 3k/f after the first block on C.
+
+Note that, if necessary, L can be ticked in order to validate the first header after L on C, without requiring any predictions.
+In the absence of the corresponding block body---ie because of the Header-Body Split---the assessment of subsequent headers on C will require predictions.
+Thus a prediction range of at least 3k/f ensures at least k blocks after the first header, and hence k+1 blocks in total, as necessary.
+
+### Question 2
+
+What behaviors could arise only if Chain Growth were violated?
+
+- For example, suppose R=k and the k+1th block after L on C was more than 3k/f slots after the first block after L on C.
+- In this case, the node might not be able to switch from D to C despite R <= k.
+  It can't switch to C without downloading the R+1 blocks, which it can't do without validating the R+1 headers, which it can't do without predicting more than 3k/f slots after the first header, which it might not be able to do.
+
+For example, suppose that L is the result of a block in relative slot number 7k/f - 2 within some epoch.
```

Comment:
Why did we choose `7k/ f - 2`?

Reply:
Because 7k/f - 2 is "slightly before one stability window before the end of the epoch".